src/adapter/src/catalog.rs

```diff
@@ -6556,6 +6817,56 @@ impl Catalog {
     pub fn ensure_not_reserved_role(&self, role_id: &RoleId) -> Result<(), Error> {
         self.state.ensure_not_reserved_role(role_id)
     }
+
+    pub fn ensure_not_reserved_object(
+        &self,
+        object_id: &ObjectId,
+        conn_id: ConnectionId,
+    ) -> Result<(), Error> {
+        match object_id {
+            ObjectId::Cluster(cluster_id) | ObjectId::ClusterReplica((cluster_id, _)) => {
+                if cluster_id.is_system() {
+                    let cluster = self.get_cluster(*cluster_id);
+                    Err(Error::new(ErrorKind::ReadOnlyCluster(
```

Comment:
Some of this logic is duplicated above in Op::GrantPrivilege. Should that call this function instead?

Reply:
To be honest, almost all of planning, sequencing, and the catalog commit are identical except for a few lines. I was tempted to implement `Revoke` and `Grant` as a single function with a boolean flag/enum, to avoid duplication. That's actually how PostgreSQL implements Grant/Revoke.

What do you think? Is that going too far?

Reply:
Ok, I think I just massively cleaned this up in https://github.com/MaterializeInc/materialize/pull/18827/commits/b04f54cfd9c90da404538ed41288d4064443b284

Reply:
Oh, I completely misunderstood this commit. This function is actually called in the sequencer, so I think it's safe to remove these checks from `Op::GrantPrivilege`.